### ISSUE TITLE
Unable to access Spinnaker UI on GCP instance

### ISSUE THREAD

**Issue by thecodejunkie** [spinnaker/spinnaker#3285]
### Issue Summary:
After installing, a fresh, copy of Spinnaker on my GCP account I am unable to access the UI on localhost:9000 even though I've copied the SSH tunnel setup command from the instance page

### Cloud Provider(s):
GCP

### Feature Area: 
UI

### Description:
I can't get it to work. I've created several new deployments without any result (both with a cluster connected and without)

When the instance has been up for a while, I run

```
gcloud compute ssh \
  --project=<name-of-project> \
  --zone=europe-west4-a \
  spinnaker-1 -- -L 9000:localhost:9000  -L 8084:localhost:8084
```

but as soon as it opens the tunnel, it tells me

```
bind: Address already in use
channel_setup_fwd_listener_tcpip: cannot listen to port: 9000
```

I did try to setup a Spinnaker instance last week and it worked then. There are no other deployment running (only keeping 1 at a time) and, as far as I can tell, there shouldn't be any other resources that are interfering.

No errors during the setup of Spinnaker, from the Marketplace

### Steps to Reproduce:
1. Install Spinnaker from Google Markerspace Solutions
2. Copy the SSH tunnel command and run it locally in terminal
3. Try browsing http://localhost:9000

### Additional Details:

Thanks!

**Comment by sbwsg** [comment]
This sounds like you're already running a service on your local machine at port 9000 and therefore when you try to port-forward the system can't use that port. Can you check whether you have something running on port 9000?  On Linux I run `netstat -nlp | grep 9000` and see the following response because I'm already running Deck locally:

`tcp        0      0 127.0.0.1:9000          0.0.0.0:*               LISTEN      222545/node`

**Comment by thecodejunkie** [comment]
Update

It works if I deploy Spinnaker to `europe-west1-b` (the default region for my gcloud configuration), but not if I deploy it to `europe-west4-a`.

Seem odd that it does not work, despite that the `--zone=europe-west4-a` flag is provided to the `gcloud compute ssh` command?

Both regions gives the 

```
bind: Address already in use
channel_setup_fwd_listener_tcpip: cannot listen to port: 9000
```

error, but only `europe-west1-b` actually work

**Comment by thecodejunkie** [comment]
@sbwsg I did poke around with that, but nothing turned up. Check my last update, the error message seems to be irrelevant, but the deployment region is having an effect 😕 Thanks!

**Comment by sbwsg** [comment]
OK I think I'm able to reproduce a problem here. In the `europe-west4-a` zone I can see that clouddriver is able to successfully start but in `europe-west4-b` I see an error in clouddriver's log output about a missing cats search bean.

Can you try going through the following steps and let me know what output you see?

1. SSH in to your europe-west4-b spinnaker machine
2. Run `sudo service spinnaker restart`
3. Wait for clouddriver to start up (takes around a minute for me)
4. Check for error output in the logs: `cat /var/log/spinnaker/clouddriver/clouddriver.log` and tell me what the errors say

For me, when I restart the spinnaker service in `europe-west4-a` I'm able to see clouddriver start up correctly and deck becomes available. When I restart in `europe-west4-b` I see clouddriver error out.

**Comment by sbwsg** [comment]
OK I think I might have a workaround in europe-west4-b: Run `hal config version edit --version 1.9.2` and then run `hal deploy apply`.

The deploy might tell you there's an error but, at least on my west4-b machine, the deployment succeeded and clouddriver was able to start. My port-forward then began to work and I could access Deck on port 9000 in my browser.

I think we need to update the image that we're providing through the spinnaker marketplace. It's currently deploying a flavor of 1.7 and we have updated both minor and patch versions since then.

**Comment by thecodejunkie** [comment]
I tried installing it in `europe-west4-a` again and proceeded to run `sudo service spinnaker restart` with the following output

> stop: Unknown instance:
> spinnaker start/running

After that I was able to access Deck, on port 9000, in my browser. I checked `cat /var/log/spinnaker/clouddriver/clouddriver.log` and could not see any obvious errors being thrown during startup, so I guess it managed to start it successfully.

A couple of follow-up questions

- How can a region have an impact on the startup of Spinnaker?
- I installed Spinnaker, using the Marketplace, just out of convenience. Is that considered the recommended approach on GCP or should we be looking at an alternative (which?) installation procedure if/when we decide to move forward with Spinnaker in our infrastructure?

Thank you so much for your help @sbwsg !

**Comment by thecodejunkie** [comment]
Think I stumbled in to another problem - not sure if it's related or not. However, I did setup a new instance on `europe-west1-b` (the one that worked all along), but this time I configured a cluster for it to use. After it started up I was, again, unable to connect to Deck

This is the Spinnaker setting that I configured

![image](https://user-images.githubusercontent.com/50543/45076716-70bc6c80-b0eb-11e8-9a9e-5ee48651f53d.png)

Here's the cluster part

![image](https://user-images.githubusercontent.com/50543/45076968-2daec900-b0ec-11e8-92c8-5cba59627165.png)

As you can see from blow, the cluster is running and it's in the in the same region.

![image](https://user-images.githubusercontent.com/50543/45078509-86806080-b0f0-11e8-9d45-50df79b2ba49.png)

However when I read the log file, I see the following at the bottomg

```
java.lang.IllegalStateException: org.springframework.boot.context.embedded.AnnotationConfigEmbeddedWebApplicationContext@160ac7fb has not been refreshed yet
	at org.springframework.context.support.AbstractApplicationContext.assertBeanFactoryActive(AbstractApplicationContext.java:1065) ~[spring-context-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1089) ~[spring-context-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at com.netflix.spinnaker.clouddriver.core.agent.CleanupPendingOnDemandCachesAgent.getCatsModule(CleanupPendingOnDemandCachesAgent.java:164) ~[clouddriver-core-3.1.3-20180529112818.jar:3.1.3-20180529112818]
	at com.netflix.spinnaker.clouddriver.core.agent.CleanupPendingOnDemandCachesAgent.run(CleanupPendingOnDemandCachesAgent.java:87) ~[clouddriver-core-3.1.3-20180529112818.jar:3.1.3-20180529112818]
	at com.netflix.spinnaker.cats.agent.RunnableAgent$RunnableAgentExecution.executeAgent(RunnableAgent.java:32) ~[cats-core-3.1.3-20180529112818.jar:3.1.3-20180529112818]
	at com.netflix.spinnaker.cats.redis.cluster.ClusteredAgentScheduler$AgentExecutionAction.execute(ClusteredAgentScheduler.java:272) ~[cats-redis-3.1.3-20180529112818.jar:3.1.3-20180529112818]
	at com.netflix.spinnaker.cats.redis.cluster.ClusteredAgentScheduler$AgentJob.run(ClusteredAgentScheduler.java:246) ~[cats-redis-3.1.3-20180529112818.jar:3.1.3-20180529112818]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_171]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_171]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_171]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_171]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_171]
```

I tried restarting the service and even more appeared in the log

<details>
<summary>Expand to see full error log</summary>

```
2018-09-05 07:47:50.079  WARN 3715 --- [ost-startStop-2] o.a.c.loader.WebappClassLoaderBase       : The web application [ROOT] appears to have started a thread named [spectator-gauge-polling-0] but has failed to stop it. This is very likely to create a memory leak. Stack trace of thread:
 sun.misc.Unsafe.park(Native Method)
 java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
 java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
 java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
 java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
 java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
 java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
 java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
 java.lang.Thread.run(Thread.java:748)
2018-09-05 07:47:50.141  INFO 3715 --- [           main] utoConfigurationReportLoggingInitializer :

Error starting ApplicationContext. To display the auto-configuration report re-run your application with 'debug' enabled.
2018-09-05 07:47:50.452 ERROR 3715 --- [           main] o.s.boot.SpringApplication               : Application startup failed

org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'searchController': Unsatisfied dependency expressed through field 'searchProviders'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'catsSearchProvider' defined in com.netflix.spinnaker.clouddriver.cache.CacheConfig: Unsatisfied dependency expressed through method 'catsSearchProvider' parameter 1; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'java.util.List<com.netflix.spinnaker.clouddriver.cache.SearchableProvider>' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {}
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:588) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:88) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessPropertyValues(AutowiredAnnotationBeanPostProcessor.java:366) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1264) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:553) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:483) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:761) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:867) ~[spring-context-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:543) ~[spring-context-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:122) ~[spring-boot-1.5.7.RELEASE.jar:1.5.7.RELEASE]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:693) ~[spring-boot-1.5.7.RELEASE.jar:1.5.7.RELEASE]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:360) ~[spring-boot-1.5.7.RELEASE.jar:1.5.7.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:303) ~[spring-boot-1.5.7.RELEASE.jar:1.5.7.RELEASE]
	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:134) [spring-boot-1.5.7.RELEASE.jar:1.5.7.RELEASE]
	at org.springframework.boot.builder.SpringApplicationBuilder$run$0.call(Unknown Source) [spring-boot-1.5.7.RELEASE.jar:1.5.7.RELEASE]
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48) [groovy-all-2.4.13.jar:2.4.13]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113) [groovy-all-2.4.13.jar:2.4.13]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125) [groovy-all-2.4.13.jar:2.4.13]
	at com.netflix.spinnaker.clouddriver.Main.main(Main.groovy:108) [clouddriver-web-3.1.3-20180529112818.jar:3.1.3-20180529112818]
Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'catsSearchProvider' defined in com.netflix.spinnaker.clouddriver.cache.CacheConfig: Unsatisfied dependency expressed through method 'catsSearchProvider' parameter 1; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'java.util.List<com.netflix.spinnaker.clouddriver.cache.SearchableProvider>' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {}
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:749) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:467) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1173) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1067) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:513) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:483) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:208) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.addCandidateEntry(DefaultListableBeanFactory.java:1316) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.findAutowireCandidates(DefaultListableBeanFactory.java:1282) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveMultipleBeans(DefaultListableBeanFactory.java:1180) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1096) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1066) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:585) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	... 22 common frames omitted
Caused by: org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'java.util.List<com.netflix.spinnaker.clouddriver.cache.SearchableProvider>' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {}
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.raiseNoMatchingBeanFound(DefaultListableBeanFactory.java:1493) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1104) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1066) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:835) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:741) ~[spring-beans-4.3.11.RELEASE.jar:4.3.11.RELEASE]
	... 38 common frames omitted
```
</details>

Any and all advice is appreciated. If you think this can be resolved, but installing a newer version using an alternative installation procedure, please let me know 😄

**Comment by sbwsg** [comment]
This is exactly the same error that I saw when deploying the image on a test VM, `Error creating bean with name 'searchController'`.

> How can a region have an impact on the startup of Spinnaker?

I'm not convinced yet that this is a zone problem since you've experienced it both in europe-west-a and europe-west-b. My assumption is that there is some kind of dependency resolution problem in the marketplace image that will be fixed when we publish a new version of that image.

> I installed Spinnaker, using the Marketplace, just out of convenience. Is that considered the recommended approach on GCP or should we be looking at an alternative (which?) installation procedure if/when we decide to move forward with Spinnaker in our infrastructure?

It depends a bit on the number of applications and pipelines you're going to need to support but I personally find the easiest route on GCP is to create a Debian VM and then run through the following steps:

1. SSH into the VM with port-forward for 9000 and 8084
2. [Install Halyard](https://www.spinnaker.io/setup/install/halyard/)
    1. `curl -O https://raw.githubusercontent.com/spinnaker/halyard/master/install/debian/InstallHalyard.sh`
    2. `sudo bash ./InstallHalyard.sh`
3. [Setup GCS storage for Front50 to save Pipelines](https://www.spinnaker.io/setup/install/storage/gcs/)
4. [Choose and configure the cloud providers you'll deploy to](https://www.spinnaker.io/setup/install/providers/)
5. Run `hal deploy apply`
6. Finally it's wise, once you're considering Spinnaker for production usage, to [externalize Redis](https://www.spinnaker.io/setup/productionize/caching/externalize-redis/) and / or [spend some cycles configuring it for Spinnaker](https://www.spinnaker.io/setup/productionize/caching/configure-redis-usage/).

> Any and all advice is appreciated. If you think this can be resolved, but installing a newer version using an alternative installation procedure, please let me know

We're going to upgrade the marketplace image and that should resolve these issues. Before then you can try the following steps:

1. SSH into your broken Spinnaker VM
2. Run the command `hal config version edit --version 1.9.2`
3. Run `hal deploy apply`

You might see errors during the last step but, at least for me, this upgrades the Spinnaker version and gets Deck successfully running.

**Comment by thecodejunkie** [comment]
Hi

Finally got time to get back to this. I ran the commands you suggested and ended up with the following error message

```
+ Get current deployment
  Success
- Prep deployment
  Failure
Problems in halconfig:
- WARNING There is a newer version of Halyard available (1.9.1),
  please update when possible
? Run 'sudo apt-get update && sudo apt-get install
  spinnaker-halyard -y' to upgrade

Problems in default.provider.kubernetes.my-kubernetes-account:
- WARNING You have not specified a Kubernetes context in your
  halconfig, Spinnaker will use
  "<redacted>" instead.
? We recommend explicitly setting a context in your halconfig, to
  ensure changes to your kubeconfig won't break your deployment.
? Options include:
  - <redacted>

! ERROR Unable to communicate with your Kubernetes cluster: Failure
  executing: GET at: https://146.148.7.86/api/v1/namespaces. Message: Forbidden!
  User <redacted> doesn't have
  permission. namespaces is forbidden: User "client" cannot list namespaces at the
  cluster scope: Unknown user "client"..
? Unable to authenticate with your Kubernetes cluster. Try using
  kubectl to verify your credentials.

- Failed to prep Spinnaker deployment
```

Not sure if this is related to my other problem, or just a configuration issue on my end? I've been following one of the guides on spinnaker.io https://www.spinnaker.io/guides/tutorials/codelabs/gcp-kubernetes-source-to-prod/ and I've configured the cluster and spinnaker as mentioned

I kept the `Kubernetes Account Name` as the default, just like in the tutorial, i.e `default.provider.kubernetes.my-kubernetes-account`

![image](https://user-images.githubusercontent.com/50543/45076968-2daec900-b0ec-11e8-92c8-5cba59627165.png)

**Comment by sathishkumar64** [comment]
Hi,

I have followed based on their document(which is following).

https://www.spinnaker.io/guides/tutorials/codelabs/gcp-kubernetes-source-to-prod/

But page is not loading, still in home page only.

![Capture](https://user-images.githubusercontent.com/20319693/57214691-5df69080-7008-11e9-99a3-a8db26ea1ad8.PNG)

Please suggest me what im doing wrong.
